src/java.base/share/classes/sun/security/provider/ML_DSA_Provider.java

```diff
@@ -0,0 +1,160 @@
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package sun.security.provider;
+
+import sun.security.jca.JCAUtil;
+import java.security.*;
+import java.security.SecureRandom;
+import java.util.Arrays;
+
+public class ML_DSA_Provider {
+
+    static int name2int(String name) {
+        if (name.endsWith("44")) return 2;
+        else if (name.endsWith("65")) return 3;
+        else if (name.endsWith("87")) return 5;
+        else throw new ProviderException();
+    }
+
+    public static class KPG extends NamedKeyPairGenerator {
+        public KPG() {
+            // ML-DSA-65 is default
+            super("ML-DSA", "ML-DSA-65", "ML-DSA-44", "ML-DSA-87");
+        }
+
+        public KPG(String pname) {
+            super("ML-DSA", pname);
+        }
+
+        @Override
+        public byte[][] implGenerateKeyPair(String name, SecureRandom sr) {
+            byte[] seed = new byte[32];
+            var r = sr != null ? sr : JCAUtil.getDefSecureRandom();
+            r.nextBytes(seed);
+            ML_DSA mlDsa = new ML_DSA(name2int(name));
+            ML_DSA.ML_DSA_KeyPair kp = mlDsa.generateKeyPairInternal(seed);
+            try {
+                return new byte[][]{
+                        mlDsa.pkEncode(kp.publicKey()),
+                        mlDsa.skEncode(kp.privateKey())};
+            } finally {
+                kp.privateKey().destroy();
+                Arrays.fill(seed, (byte)0);
+            }
+        }
+    }
+
+    public static class KPG2 extends KPG {
+        public KPG2() {
+            super("ML-DSA-44");
+        }
+    }
+
+    public static class KPG3 extends KPG {
+        public KPG3() {
+            super("ML-DSA-65");
+        }
+    }
+
+    public static class KPG5 extends KPG {
+        public KPG5() {
+            super("ML-DSA-87");
+        }
+    }
+
+    public static class KF extends NamedKeyFactory {
+        public KF() {
+            super("ML-DSA", "ML-DSA-44", "ML-DSA-65", "ML-DSA-87");
+        }
+        public KF(String name) {
+            super("ML-DSA", name);
+        }
+    }
+
+    public static class KF2 extends KF {
+        public KF2() {
+            super("ML-DSA-44");
+        }
+    }
+
+    public static class KF3 extends KF {
+        public KF3() {
+            super("ML-DSA-65");
+        }
+    }
+
+    public static class KF5 extends KF {
+        public KF5() {
+            super("ML-DSA-87");
+        }
+    }
+
+    // TODO: check key in initSign and initVerify?
+    public static class SIG extends NamedSignature {
+        public SIG() {
+            super("ML-DSA", "ML-DSA-44", "ML-DSA-65", "ML-DSA-87");
```

Comment:
Add a comment with the default as you do for KPG.

Reply:
See above.